presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+        }
+    }
+
+    private void setShared()
+    {
+        checkReadable();
+        state = State.SHARED_OPERATION_BUFFERED;
+    }
+
+    private void setExclusive(ExclusiveOperation exclusiveOperation)
+    {
+        if (state != State.EMPTY) {
+            throw new PrestoException(StandardErrorCode.NOT_SUPPORTED, "Unsupported combination of operations in a single transaction");
+        }
+        state = State.EXCLUSIVE_OPERATION_BUFFERED;
+        bufferedExclusiveOperation = exclusiveOperation;
+    }
+
+    private void checkNoPartitionAction(String databaseName, String tableName)
+    {
+        Map<List<String>, Action<PartitionAndMore>> partitionActionsOfTable = partitionActions.computeIfAbsent(new SchemaTableName(databaseName, tableName), k -> new HashMap<>());
+        if (!partitionActionsOfTable.isEmpty()) {
+            throw new PrestoException(NOT_SUPPORTED, "Making schema changes to a table/view whose partitions were modified in the same transaction is not supported");
+        }
+    }
+
+    private static boolean isSameOrParent(Path parent, Path child)
+    {
+        int parentDepth = parent.depth();
+        int childDepth = child.depth();
+        if (parentDepth > childDepth) {
+            return false;
+        }
+        for (int i = childDepth; i > parentDepth; i--) {
+            child = child.getParent();
+        }
+        return parent.equals(child);
+    }
+
+    private void logCleanupFailure(String format, Object... args)
+    {
+        if (throwOnCleanupFailure) {
+            throw new RuntimeException(format(format, args));
+        }
+        log.warn(format, args);
+    }
+
+    private void logCleanupFailure(Throwable t, String format, Object... args)
+    {
+        if (throwOnCleanupFailure) {
+            throw new RuntimeException(format(format, args), t);
+        }
+        log.warn(t, format, args);
+    }
+
+    private static void waitIgnoreThrowables(List<CompletableFuture<?>> futures)
+    {
+        for (CompletableFuture<?> future : futures) {
+            try {
+                future.get();
+            }
+            catch (Throwable t) {
+                // ignore
+            }
+        }
+    }
+
+    private static void asyncRename(
+            HdfsEnvironment hdfsEnvironment,
+            Executor executor,
+            AtomicBoolean cancelled,
+            List<CompletableFuture<?>> fileRenameFutures,
+            String user,
+            Path currentPath,
+            Path targetPath,
+            List<String> fileNames)
+    {
+        FileSystem fileSystem;
+        try {
+            fileSystem = hdfsEnvironment.getFileSystem(user, currentPath);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error moving INSERT_EXISTING data to final location. Error listing directory %s", currentPath), e);
```

Comment:
I don't think the error message should contain `INSERT_EXISTING` since that is an internal detail